storage/execout/file.go

```diff
@@ -2,213 +2,332 @@ package execout
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"io"
 	"path"
 	"sort"
-	"strconv"
-	"sync"
 
-	"connectrpc.com/connect"
 	pboutput "github.com/streamingfast/substreams/storage/execout/pb"
 	"github.com/streamingfast/substreams/storage/execout/streamproto"
 
 	"go.uber.org/zap/zapcore"
 
-	"github.com/streamingfast/derr"
 	"github.com/streamingfast/dstore"
 	"go.uber.org/zap"
 
 	"github.com/streamingfast/substreams/block"
 	pbsubstreams "github.com/streamingfast/substreams/pb/sf/substreams/v1"
 )
 
-// A File in `execout` stores, for a given module (with a given hash), the outputs of module execution
-// for _multiple blocks_, based on their block ID.
-type File struct {
-	sync.RWMutex
-	*block.Range
+type FileReader interface {
+	ReadNext() (*pboutput.Item, error)
+	Get(ctx context.Context, blockNumber uint64) (payload []byte, found bool, err error)
+	ModuleName() string
+	Filename() string
+}
 
-	ModuleName string
-	Kv         map[string]*pboutput.Item
-	store      dstore.Store
-	logger     *zap.Logger
-	loaded     bool
+type fileReader struct {
+	*File
+	reader       io.ReadCloser
+	lastReadItem *pboutput.Item
+	complete     bool
+}
 
-	writingFile   *io.PipeWriter
-	writeError    chan error
-	deletedBefore *uint64
+func NewFileReader(ctx context.Context, store dstore.Store, logger *zap.Logger, rng *block.Range, moduleName string) (FileReader, error) {
```

Comment:
```suggestion
func OpenFileReader(ctx context.Context, store dstore.Store, logger *zap.Logger, rng *block.Range, moduleName string) (FileReader, error) {
```

The name would be more closely aligned with `os.OpenFile` for example. Otherwise, maybe separating the `New` and the `Open` on the struct would be better.